File: solution/894. All Possible Full Binary Trees/894.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> allPossibleFBT(int n) {
        // if (n % 2 == 0) return {};
        if (n == 1) return { new TreeNode(0) };
        
        vector<TreeNode*> res;
        for (int i = 1; i < n - 1; i += 2) {
            vector<TreeNode*> leftNodes = allPossibleFBT(i);
            vector<TreeNode*> rightNodes = allPossibleFBT(n - i - 1);
            for (auto& l : leftNodes) {
                for (auto& r : rightNodes) {
                    TreeNode* node = new TreeNode(0);
                    node->left = l;
                    node->right = r;
                    res.push_back(node);
                }
            }
        }
        return res;
    }
};


// 0 1 2 3 4 5 6
// x x x x x x x
//   ^   ^   ^

File: solution/894. All Possible Full Binary Trees/894.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def allPossibleFBT(self, n: int) -> List[Optional[TreeNode]]:
        if n % 2 == 0:
            return []
        if n == 1:
            return [ TreeNode() ]

        res = []
        for i in range(1, n - 1, 2):
            l_nodes = self.allPossibleFBT(i)
            r_nodes = self.allPossibleFBT(n - 1 - i)
            for l in l_nodes:
                for r in r_nodes:
                    node = TreeNode()
                    node.left = l
                    node.right = r
                    res.append(node)
        return res
